wrote redmefile (#16)
mmm1513/pyminio: README.md

```bash
pip install pyminio
```
or
```bash
git clone https://github.com/mmm1513/pyminio.git
cd pyminio
python setup.py install
```

## Setting up Pyminio

Firstly you need to set up your  [Minio Docker](https://hub.docker.com/r/minio/minio/), and acquire an ENDPOINT (URL), ACCESS_KEY, and a SECRET_KEY.

- If you want to add your own minio object you can pass it in the constructor like so:

    Install python's [Minio](https://docs.min.io/docs/python-client-quickstart-guide.html) module.

    ```python
    from minio import Minio
    from pyminio import Pyminio

    minio_obj = Minio(
        endpoint='<your-minio-endpoint>',  # e.g. "localhost:9000/"
        access_key='<your-minio-access-key>',
        secret_key='<your-minio-secret-key>'
    )
    pyminio_client = Pyminio(minio_obj=minio_obj)
    ```

- If you don't want to handle with minio, you can do this instead:

    ```python
    from pyminio import Pyminio

    pyminio_client = Pyminio.from_credentials(
        endpoint='<your-minio-endpoint>',  # e.g. "localhost:9000/"
        access_key='<your-minio-access-key>',
        secret_key='<your-minio-secret-key>'
    )
    ```

## Usage

> - [mkdirs](#self.mkdirs(...))

Reviewer: you can use 
```
In header name:
# <a name="mkdir"></a>A Heading
here:
[mkdirs](#mkdir)
```

Author: d

- [listdir](#self.listdir(...))
- [exists](#self.exists(...))
- [isdir](#self.isdir(...))
- [truncate](#self.truncate())
- [rmdir](#self.rmdir(...))
- [rm](#self.rm(...))
- [cp](#self.cp(...))
- [mv](#self.mv(...))
- [get](#self.get(...))
- [get_last_object](#self.get_last_object(...))
- [put_data](#self.put_data(...))
- [put_file](#self.put_file(...))

> ### self.mkdirs(...)

Reviewer: I mean you to fill in the whole signature.
... doesnt give me anything

Author: d

# Pyminio

> Minio is a python client wrapped like os to control minio server.

Reviewer 2: ```suggestion
Minio is a python client wrapped like the `os` module to control minio server.
```

> Minio is a python client wrapped like os to control minio server.

Reviewer 2: After this line, you can elaborate and explain why you wrote it, how it relates to the original client and why it's better than the original client (only in general, you don't need to get into the details)

[![PyPI](https://img.shields.io/pypi/v/pyminio?color=blue&label=pypi%20version)]()
[![PyPI](https://img.shields.io/pypi/pyversions/pyminio.svg?style=plastic)]()
[![Downloads](https://pepy.tech/badge/pyminio)](https://pepy.tech/project/pyminio)

## Content
1. [Installation](#Installation)
2. [Setting up Pyminio](#Setting-up-Pyminio)
3. [Usage](#Usage)
   - [mkdirs](#mkdirs)
   - [listdir](#listdir)
   - [exists](#exists)
   - [isdir](#isdir)
   - [truncate](#truncate)
   - [rmdir](#rmdir)
   - [rm](#rm)
   - [cp](#cp)
   - [mv](#mv)
   - [get](#get)
   - [get_last_object](#get_last_object)
   - [put_data](#put_data)

>    - [put_file](#put_file)

Reviewer 2: It's unconventional to put so much sub-sections in the README
either remove it from the Table of Contents or put this hierarchy on readthedocs (either way remove those sub-sections from README)

Author: moved it to usage

4. [Contribute](#Contribute)

## Installation
Use the package manager [pip](https://pypi.org/project/pyminio/) to install pyminio.
```bash
pip install pyminio

> ```

Reviewer 2: add how to install from source
lots of people love it for some reason

> ## Setting up Pyminio:

Reviewer: no need for : remove everywhere in titlts

> In case you want to add your own minio object you can pass it in the constructor like so:

Reviewer 2: FIrst, start with a request to setup minio itself (add a link to the relevant minio's installation guide)
Then, tell them what they need to acquire from that step (URL, ACCESS_KEY, SECRET_KEY)
Then comes the usage of Pyminio that you already wrote

```python
from minio import Minio

> from pyminio import Pyminio

Reviewer 2: add comments for each line, so people can understand what happens here

ENDPOINT = environ.get('MINIO_CONNECTION')
ACCESS_KEY = environ.get('MINIO_ACCESS_KEY')
SECRET_KEY = environ.get('MINIO_SECRET_KEY')

minio_obj = Minio(

>             endpoint=self.ENDPOINT,

Reviewer 2: append comment of how it looks
```suggestion
            endpoint=self.ENDPOINT,  # e.g. minio://localhost:38989
```
I don't  really know how this URL looks, I just gave an example above (so change accordingly to schema and default port)

same for ACCESS_KEY and SECRET_KEY if they tend to have some structure (or a fixed length)

>             endpoint=self.ENDPOINT,

Reviewer 3: remove `self`

            access_key=self.ACCESS_KEY,
            secret_key=self.SECRET_KEY
        )
pyminio_client = Pyminio(minio_obj=minio_obj)
```

> if you dont want to handle with minio, you cand do this instead:

Reviewer 2: *can
*If
*don't

Reviewer: don't resolve before pushing.
I don't see it still

Author: next pr xD

> if you dont want to handle with minio, you cand do this instead:

Reviewer 2: Maybe change the cases
Usually, people start with the simplest approach available, only after they advance to the more customized approach

Author: for me this is the simplest approach, tho now its shorter

```python
from pyminio import Pyminio

ENDPOINT = environ.get('MINIO_CONNECTION')
ACCESS_KEY = environ.get('MINIO_ACCESS_KEY')

> SECRET_KEY = environ.get('MINIO_SECRET_KEY')

Reviewer 2: BTW, you can also make a factory that fetches it from env-vars, like:
```python
pyminio_client = Pyminio.from_env()
```

Author: i removed that env var thing

pyminio_client = Pyminio.from_credentials(

>     endpoint=self.ENDPOINT,

Reviewer 3: remove self

Author: removed

    access_key=self.ACCESS_KEY,
    secret_key=self.SECRET_KEY
)
```

> You can storage your minio credentials in environment varibles like in these examples or pass them to Pyminio in any other way.

Reviewer 2: *store
*variables
maybe:
```suggestion
It's recommended that you store Minio's access info in environment variables, to make your code more portable.
```

Author: removed

## Usage:

### mkdirs

> Pyminio.mkdirs will create the given full path if not exists like linux's 'mkdir -p'.

Reviewer: `pyminio.mkdir`
use backtick in places you referance code references

Author: d

> This function must get a directory path or it will raise a ValueError.

Reviewer 2: You mean it will raise if given `/foo/bar/baz` instead of `/foor/bar/baz/?
Then you must include this example below the one given, and be VERY explicit about it in the error message

Author: d

```python
>>> pyminio_client.mkdirs('/foo/bar/baz/')
```

> ### listdir

Reviewer: i actually prefer seeing the signature:
self.listdir(....)

Author: d

> Pyminio.listdir will return the directory's content as a list of directorys and files name. Works like os's 'listdir'.

Reviewer 2: Maybe
```suggestion
Pyminio.listdir will return the bucket/folder content as a list of directories and file names. Works like os's 'listdir'.
```
Note typos fixed above as well

Author: only directory's i dont want the user to acknowledge buckets

This function must get a directory path or it will raise a ValueError.

```python
>>> pyminio_client.listdir('/foo/bar/baz/')

> ['file_name_1', 'file_name_2', 'directory_name/']

Reviewer: is it returns a list? or a tuple?
make sure its a tuple :)

Author: d

```

> There is an option to use the only_files flag to get only files from listdir.

Reviewer 2: maybe add `only_dirs` for the symmetry

Author: d

```python

> >>> pyminio_client.listdir('/foo/bar/baz/', only_files=True)

Reviewer: change to `files_only` kwarg :)

Author: d

['file_name_1', 'file_name_2']
```

### exists

> Pyminio.exists will return a boolean that confirm reather this path exists or not in the server. Works like os.path.exists.

Reviewer 2: *rather
format `os.path.exists` like you did above

```python

> >>> pyminio_client.exists('/foo/bar/baz/file_name_1')  # existed file

Reviewer 2: Maybe draw the tree OS in a comment and remove the comments (it will be clearer)
or change to "existing file"

Author: done

True
>>> pyminio_client.exists('/foo/bar/baz/file_name_3')  # not existed file
False
>>> pyminio_client.exists('/all/path/wrong/')  # not existed path
False
```

### isdir

> Pyminio.isdir will return True only if the given path exists and is a directory. Works like os.path.isdir

Reviewer 2: format func

```python
>>> pyminio_client.isdir('/foo/bar/baz/file_name_1')  # existed file
False
>>> pyminio_client.isdir('/foo/bar/baz/directory_name/')  # existed directory
True
>>> pyminio_client.isdir('/all/path/wrong/but/directory/')  # not existed directory
False
```

### truncate

> Pyminio.truncate will delete all minio's content.

Reviewer 2: I don't understand what it does from this explanation
also, truncate is a pretty confusing verb, so think if it fits

Reviewer: truncate actually is very known from SQL worlds.
maybe void could be a better name :\ I dont think so.
got a suggestion?

Author: changed the explanation

```python
>>> pyminio_client.truncate()
```

### rmdir

> Pyminio.rmdir will delete the specified directory. Works like linux's 'rmdir (-r)'.

Reviewer 2: `rmdir` has no option `-r`
maybe "It imitates Linux's commands `rmdir` / `rm -r`"

Author: done

> It will raise a DirectoryNotEmptyError if given directory is not empty, except if the recursive flag is on and then it will delete given directory's path recursively.

Reviewer 2: format exception
*flag is set

This function must get a directory path or it will raise a ValueError.

```python
>>> pyminio_client.rmdir('/foo/bar/baz/directory_name/')  # empty directory
>>> pyminio_client.rmdir('/foo/bar/')  # unempty directory

> DirectoryNotEmptyError: can not recursively delete unempty directory

Reviewer 2: *non-empty

>>> pyminio_client.rmdir('/foo/bar/', recursive=True)
```

### rm

> Pyminio.rm works like [rmdir](#rmdir) only that it can delete files too. Works like linux's rm (-r).

Reviewer 2: It's weird. both Linux and Python doesn't allow `rm`/`os.Remove` to delete both files and dirs

Author: linux'srm does remove files

```python
>>> pyminio_client.rm('/foo/bar/baz/file_name')
```

### cp
Pyminio.cp will copy one file or directory to given destination. Works like linux's 'cp (-r)'.

> This func can only copy recursively when the recursive flag is True. If not, it will raise a ValueError.

Reviewer 2: *func -> method
format

Author: done

#### How will the copy accure? (all directories are copied recursively in this examples)
| src path   | dst path  | dst exists | new dst      | Explain                                                                |
| ---------- | --------- | ---------- | ------------ | ---------------------------------------------------------------------- |

> | /foo/bar   | /foo/baz  |    ---     | /foo/baz     | The file's name will be changed from bar to baz.                       |

Reviewer 2: What? Then it is move and not copy
You mean that the file's content will be moved to `baz` as well?

Author: mb

> | /foo1/bar  | /foo2/    |    True    | /foo/bar     | The file will be copied to '/foo2/'                                    |

Reviewer 2: new dst is `/foo2/bar` no?
I'm really confused, let's  talk about what happens here

Author: mb

| /foo/bar/  | /foo/     |    True    | /foo/        | The content of '/foo/bar/' will be copied to '/foo/'                   |
| /foo1/bar/ | /foo2/    |   False    | /foo2/bar/   | '/foo1/bar/' will be copied recursively to '/foo2/bar/'                |
| /foo1/bar/ | /foo2/baz |    ---     |     ---      | ValueError will be raised in attempting to copy directory in to a file |

```python
>>> pyminio_client.cp('/foo/bar', '/foo/baz')
>>> pyminio_client.cp('/foo1/bar', '/foo2/')
>>> pyminio_client.cp('/foo/bar/', '/foo/', recursive=True)
>>> pyminio_client.cp('/foo1/bar/', '/foo2/', recursive=True)
>>> pyminio_client.cp('/foo1/bar/', '/foo2/baz', recursive=True)
ValueError: can not activate this method from directory to a file.
```

### mv

> Pyminio.mv works like [cp](#cp) only that it remove the source after complete the transfare. Works like linux's 'mv (-r)'.

Reviewer 2: Pyminio.mv -> the `mv` method
*only that it removes the source after the transfer has been completed
No `-r` in `mv`

This func can only move recursively when the recursive flag is True. If not, it will raise a ValueError.

```python
>>> pyminio_client.mv('/foo/bar/', '/foo/baz/')
```

### get

> Pyminio.get return an object from given path. This object will return as a pyminio.File object or an pyminio.Folder object, that both inherit from pyminio.ObjectData

Reviewer 2: *will be returned

This objects will contain metadata, their path and name.

> If its a File object it will contains it's data.

Reviewer 2: ```suggestion
If it's a File object it will contain its data.
```

> If its a File object it will contains it's data.

Reviewer 2: I think it's a mistake to print or save the file's content

Author: i just return the data that was asked, do you think that in get to an obj i should not return its data??

```python
>>> pyminio_client.get('/foo/bar/baz')

> File(name='baz', full_path='/foo/bar/baz', metadata=AttrDict({'is_dir': False, 'last_modified': time.struct_time(...), 'size': ..., 'content-type': ...}), data=...)

Reviewer 3: line is too long
add enters to remove the scoller

```

### get_last_object
Pyminio.get_last_object will return the last modified object inside a given directory.

This function must get a directory path or it will raise a ValueError.

```python
>>> pyminio_client.get_last_object('/foo/bar/')
File(name='baz', full_path='/foo/bar/baz', metadata=AttrDict({'is_dir': False, 'last_modified': time.struct_time(...), 'size': ..., 'content-type': ...}), data=...)
```

### put_data
Pyminio.put_data gets a path, data in bytes, and some metadata, and create an object inside the given path.

```python
>>> data = b'test'
>>> metadata = {'Pyminio-is': 'Awesome'}
>>> pyminio_client.put_data(path='/foo/bar/baz', data=data, metadata=metadata)
```

### put_file

> Pyminio.put_file works like [put_data](#put_data) only that instead of data it gets a path to a file in you computer. Then it will copy this file to the given location.

Reviewer 2: do you support file-like objects?

Author: no, is it important?

```python
>>> metadata = {'Pyminio-is': 'Awesome'}

> >>> pyminio_client.put_data(path='/foo/bar/baz', file_path='/mnt/some_file', metadata=metadata)

Reviewer 2: `path` and `file_path` is very confusing
you need to make sure the user understands which is the source and which is the target
make sure the source is the first one (just like Linux)
also, you may be better using keyword-only arguments, like so:
```python
def put_data(*, source_path, dest_path, metadata)
```

Author: d

```

## Contribute

All contributions are welcome:

> - Read the [issues](https://github.com/mmm1513/pyminio/issues), Fork the [project](https://github.com/mmm1513/pyminio) and do a Pull Request.

Reviewer: and create a new Pull Request

> - Request a new topic creating a `New issue` with the `enhancement` tag.

Reviewer 2: ```suggestion
- Request a new feature creating a `New issue` with the `enhancement` tag.
```

> - Request a new topic creating a `New issue` with the `enhancement` tag.

Reviewer: actually you are the one who should label issues IMO

Author: k

> - Find any kind of errors in the code and create a `New issue` with the details and the `bug enhancement` or fork the project and do a Pull Request.

Reviewer 2: there's a `bug enhancement` tag? that's counter-intuitive :)

Author: removed

- Suggest a better or more pythonic way for existing examples.

### Work environment:

> After forking the project and installing the dependencies,

Reviewer: explain how to install deps not everyone knows

Author: d

> download the [minio docker](https://hub.docker.com/r/minio/minio/) and start an instance in your computer for development and testing.

Reviewer: add a detailed way to start developing.
this is not enough

Reviewer: docker run -p

> #### Dont forget to write tests, and to run all the tests before making a pull request.

Reviewer 2: *Don't

> #### Dont forget to write tests, and to run all the tests before making a pull request.

Reviewer 2: add tox, it will make it easier for contributers

Author: opened issue #17

> #### Dont forget to write tests, and to run all the tests before making a pull request.

Reviewer 3: Add the command line to run the tests

Author: d

> #### Dont forget to write tests, and to run all the tests before making a pull request.

Reviewer: add PULL_REQUEST_TEMPLATE.md (in root repo) file. it will help alot with showing people what are the required steps
https://docs.github.com/en/free-pro-team@latest/github/building-a-strong-community/creating-a-pull-request-template-for-your-repository

        access_key='<your-minio-access-key>',
        secret_key='<your-minio-secret-key>'
    )
    pyminio_client = Pyminio(minio_obj=minio_obj)
    ```

- If you don't want to handle with minio, you can do this instead:

    ```python
    from pyminio import Pyminio

    pyminio_client = Pyminio.from_credentials(
        endpoint='<your-minio-endpoint>',  # e.g. "localhost:9000/"
        access_key='<your-minio-access-key>',
        secret_key='<your-minio-secret-key>'
    )
    ```

## Usage
- [mkdirs](#self.mkdirs(...))
- [listdir](#self.listdir(...))
- [exists](#self.exists(...))
- [isdir](#self.isdir(...))
- [truncate](#self.truncate())
- [rmdir](#self.rmdir(...))
- [rm](#self.rm(...))
- [cp](#self.cp(...))
- [mv](#self.mv(...))
- [get](#self.get(...))
- [get_last_object](#self.get_last_object(...))
- [put_data](#self.put_data(...))
- [put_file](#self.put_file(...))

### self.mkdirs(...)
`Pyminio.mkdirs` will create the given full path if not exists like linux's `mkdir -p`.

This method must get a directory path or it will raise a ValueError.

```python
>>> pyminio_client.mkdirs('/foo/bar/baz/')
>>> pyminio_client.mkdirs('/foo/bar/baz')

> ValueError /foo/bar/baz is not a valid directory path. must be absolute and end with /

Reviewer: add {path!r} to the string formatting of this error

Author: d

```

### self.listdir(...)
`Pyminio.listdir` will return the directory's content as a tuple of directories and file names. Works like os's `listdir`.

This method must get a directory path or it will raise a ValueError.

```python
>>> pyminio_client.listdir('/foo/bar/baz/')
('file_name_1', 'file_name_2', 'directory_name/')
```

There is an option to use the files_only flag to get only files and dirs_only to get only directories from listdir.

```python
>>> pyminio_client.listdir('/foo/bar/baz/', files_only=True)
('file_name_1', 'file_name_2')
>>> pyminio_client.listdir('/foo/bar/baz/', dirs_only=True)

> ('directory_name/')

Reviewer: you are missing , (tuple with length 1)

Author: d

```

### self.put_file(...)
`Pyminio.put_file` works like [put_data](#put_data) only that instead of data it gets a path to a file in you computer. Then it will copy this file to the given location.

```python
>>> metadata = {'Pyminio-is': 'Awesome'}
>>> pyminio_client.put_file(to_path='/foo/bar/baz', file_path='/mnt/some_file', metadata=metadata)
```

## Contribute

All contributions are welcome:

- Read the [issues](https://github.com/mmm1513/pyminio/issues), Fork the [project](https://github.com/mmm1513/pyminio) and create a new Pull Request.
- Request a new feature creating a `New issue` with the `enhancement` tag.
- Find any kind of errors in the code and create a `New issue` with the details, or fork the project and do a Pull Request.
- Suggest a better or more pythonic way for existing examples.

### Work environment

After forking the project and installing the dependencies, (like specified in the [installations](#Installation) in part 2)
download the [minio docker](https://hub.docker.com/r/minio/minio/) and start an instance in your computer for development and testing.

to run the tests run:
```bash

> pytest -v tests/tests.py

Reviewer: only
```
pytest
```
is also working :)

Author: d
